packages/gatsby/src/utils/create-node-id.ts

```diff
@@ -16,9 +34,40 @@ export function createNodeId(id: string | number, namespace: string): string {
     id = id.toString()
   } else if (typeof id !== `string`) {
     report.panic(
-      `Parameter passed to createNodeId must be a String or Number (got ${typeof id})`
+      `The \`id\` parameter passed to createNodeId must be a String or Number (got ${typeof id})`
+    )
+  } else if (typeof namespace !== `string`) {
+    report.panic(
+      `The \`namespace\` parameter passed to createNodeId must be a String (got ${typeof namespace})`
     )
   }
 
-  return uuidv5(id, uuidv5(namespace, seedConstant))
+  // Note: we can use the idCache for namespace as well since the key for ids will include the namespace
+  let nsHash = idCache.get(namespace)
+  if (!nsHash) {
+    nsHash = uuidv5(namespace, seedConstant)
+    idCache.set(namespace, nsHash)
+  }
```

Comment:
unlikely, but sharing `idCache` for namespaces and nodes might cause collisions due to how keys are generated:
 - `${namespace}` for namespace UUID
 - `${namespace}_${id}` for node UUID

With example (once again, unlikely) scenario:
Namespace UUID for `gatsby-plugin-foo_bar` and node with `bar` id in `gastsy-plugin-foo` namespace would generate same key for `idCache` - maybe just need to separate tho?

Even then tho, there is still risk of same key for nodes themselves - it just need per namespace lookup maps (instead of concatenating `namespace` and `id`) or maybe just more unlikely joiner string (so not just `_`, but something much more uncommon)? Example for above potential issue: namespace `A` and id `foo_bar` generates same key as namespace `A_foo` and id `bar`

Reply:
Fair point. Let's hope the double layer maps don't moot the win, or make it worse for certain cases.

Reply:
Silly (well maybe not) idea - add `namespace.length` (or `id.length`) to generated key to so there is actually difference when generating key when there previously was room for collisions?

So:
 - namespace `A`, node id: `foo_bar` generates `A_foo_bar_1`
 - namespace `A_foo`, node id: `bar` generates `A_foo_bar_5`

Not sure if addresses it enough or double level lookup are still the way to go

Reply:
Implemented a double layered map